cmd/od-main.go

```diff
@@ -0,0 +1,235 @@
+func (o odMessage) JSON() string {
+	copyMessageBytes, e := json.MarshalIndent(o, "", " ")
+	fatalIf(probe.NewError(e), "Unable to marshal into JSON.")
+
+	return string(copyMessageBytes)
+}
+
+func setOdSizes(args map[string]string, odURLs URLs) (combinedSize int64, partSize uint64, parts int, e error) {
+	// If parts not specified, set to 0, else scan for integer.
+	p := args["parts"]
+	if p == "" {
+		parts = 0
+	} else {
+		parts, e = strconv.Atoi(p)
+		if e != nil {
+			return 0, 0, 0, e
+		}
+	}
+
+	s := args["size"]
+	if parts < 1 && s == "" {
+		if parts == 0 {
+			return 0, 0, 0, fmt.Errorf("either parts or size must be specified")
+		}
+		return 0, 0, 0, fmt.Errorf("parts must be at least 1 or size must be specified")
+	}
+
+	// If size is not specified, calculate the size of each part and upload full file.
+	filesize := odURLs.SourceContent.Size
+	if s == "" {
+		combinedSize = filesize
+		partSize = uint64(math.Ceil(float64(combinedSize) / float64(parts)))
+		return combinedSize, partSize, parts, nil
+	}
+
+	partSize, e = humanize.ParseBytes(s)
+	if e != nil {
+		return 0, 0, 0, e
+	}
+
+	// If parts is not specified, calculate number of parts and upload full file.
+	if parts < 1 {
+		combinedSize = filesize
+		parts = int(math.Ceil(float64(combinedSize) / float64(partSize)))
+		return combinedSize, partSize, parts, nil
+	}
+
+	combinedSize = int64(partSize * uint64(parts))
+	if filesize < combinedSize {
+		return 0, 0, 0, fmt.Errorf("size of source file is smaller than the combined size of parts")
+	}
+	return combinedSize, partSize, parts, nil
+}
+
+func getOdUrls(args map[string]string, ctx context.Context) (odURLs URLs, e error) {
+	inFile := args["if"]
+	outFile := args["of"]
+
+	// Placeholder encryption key database
+	var encKeyDB map[string][]prefixSSEPair
+
+	// Check if outFile is a folder or a file.
+	opts := prepareCopyURLsOpts{
+		sourceURLs: []string{inFile},
+		targetURL:  outFile,
+	}
+	odType, _, err := guessCopyURLType(ctx, opts)
+	fatalIf(err.Trace(), "Unable to guess copy URL type")
+
+	// Get content of inFile, set up URLs.
+	switch odType {
+	case copyURLsTypeA:
+		odURLs = prepareCopyURLsTypeA(ctx, inFile, "", outFile, encKeyDB)
+	case copyURLsTypeB:
+		odURLs = prepareCopyURLsTypeB(ctx, inFile, "", outFile, encKeyDB)
+	default:
+		return URLs{}, fmt.Errorf("invalid source path %s, source cannot be a directory", inFile)
+	}
+
+	// Check if source exists.
+	if odURLs.SourceContent == nil {
+		return URLs{}, fmt.Errorf("invalid source path %s", inFile)
+	}
+
+	// Check if target alias is valid.
+	if odURLs.TargetAlias == "" {
+		return URLs{}, fmt.Errorf("invalid target path %s", outFile)
+	}
+
+	return odURLs, nil
+}
+
+func mainOD(cliCtx *cli.Context) error {
+	ctx, cancelCopy := context.WithCancel(globalContext)
+	defer cancelCopy()
+
+	if !cliCtx.Args().Present() {
+		cli.ShowCommandHelpAndExit(cliCtx, "mb", 1) // last argument is exit code
+	}
+
+	mapArgs, err := parseKVArgs(strings.Join(cliCtx.Args(), ","))
```

Comment:
Use a more typed representation of parsed values using, map is buggy and can lead to silent bugs hard to debug.

Reply:
Would using madmin KVS make sense for this?